[FEAT] ProductMenu 기능 구현 - 장세훈

JSeHoone opened the pull request:
### 목적
카페에서 판매중, 판매 예정 등의 상품을 관리할 수 있는 서비스를 제공한다.

### 기능
[메뉴 관리 플로우](https://www.notion.so/ohgiraffers/27e649136c1180b5a0bcc21d3dc696a3?source=copy_link)
 1. 메뉴판 제공
 2. 메뉴판 상태관리
 3. 신메뉴 등록 및 기존 메뉴 삭제

### PR 템플릿
📚 학습 내용
- 기존에 InMemoryData에 있는 정보를 가져와서 패키지 안의 Database에 옮겼습니다.
- 패키지 내부의 DataBase로 데이터를 관리했는데, 
- 메인 페이지로 돌아갈 때 마다 초기화 되는 문제가 있어서 싱글톤이라는 것을 학습해 적용했습니다. 

🔍 주요 변경사항
1. 공통 파일의 주요 변경 사항으로 ManagerDashBoard의 ProductMenu 부분만 수정되었습니다.

📝 코드 설명
Service/menu 패키지 내의 패키지들에 대해서 설명 드리겠습니다.
- 1. domain/ : 해당 패키지는 menu를 관리하기 편하게 하기 위해 생성된 패키지 입니다.
- 2. loader/ : InMemoryData에 있는 Menu데이터를 가져와서 tables패키지에 있는 Map에 저장하기 위함입니다.
- 3. service/ : 비즈니스 로직을 수행하는 패키지로 tables패키지의 Map에 저장과 변경을 수행합니다.
- 4. tables/ : Menu데이터를 저장해두는 패키지로 Map형태로 저장되고 있고, DB의 Tabel처럼 관리하고자 이렇게 패키지명을 지었습니다.
- 5. view/ : CLI에서 보여지는 글씨와 정보들을 관리하기 위한 패키지

Service/menu/ProductMenu.java : 해당 파일은 ManagerDashBoard에서 접근시 처음으로 접근되는 파일입니다.

✅ 체크리스트

- [ ] 메뉴 관리 플로우대로 구현하였는가 ?
- [ ] 예상 가능한 오류가 없는가 ?

🤔 질문사항 (선택)
- View 부분에 중복되는 코드가 많은 것 같은데, 어떻게 하면 줄일 수 있을까요 ?
- 객체 지향적으로 개발이 되었을까요 ?..
- Enum의 이름 값을 한국어로 사용했는데, 문제가 없을까요 ?

JSeHoone committed:
feat : ProductMenu에 진입했을 때 호출 되는 메서드 정의

- run() 메서드를 수행하면 선택할 수 있도록 Scanner정의
- 1. 메뉴판을 확인하는 '메뉴 확인'
- 2. 메뉴에 대한 관리를 하는 '메뉴 관리'

JSeHoone committed:
feat : menuView 추가

- productMenu에서 1번 '메뉴판 출력'을 선택했을 때 보이 화면
- 내부적으로 InMemoryData의 Menu를 가져와서 출력함.

JSeHoone committed:
feat : menuManagerView 추가

- 메뉴 관리 뷰
- 1. 메뉴판 상태 관리뷰(메뉴의 가격 설정 및 상태(판매중, 등등))
- 2. 메뉴 등록 및 삭제뷰(신메뉴를 등록하거나 기존 메뉴를 삭제하는 뷰)

JSeHoone committed:
feat : 2번 선택 시 menuManagerView로 이동하도록 로직 수정

JSeHoone committed:
refactor : Class파일명 변경

- menuView -> MenuBoardView 로 변경
- menuManager -> MenuManagerView로 변경
- ProductMenu -> 변경된 Class파일명으로 인한 수정

JSeHoone committed:
feat : 메뉴 관리 페이지에서 접근하는 2개의 페이지 추가

- 1. MenuBoardSettingView : 메뉴판 상태관리 페이지
- 2. AddAndDeleteMenuView : 신메뉴 추가 및 삭제 페이지

JSeHoone committed:
refactor : MenuMangerView에 다음 페이지로 넘어가도록 수정

- 1. 메뉴판 관리 페이지로 넘어가도록 수정
- 2. 신메뉴 등록 및 삭제 페이지로 넘어가도록 수정

JSeHoone committed:
feat : Menu관리를 위한 domain 및 table 패키지 추가

- domain/MenuEntity : 메뉴를 관리하기 위한 엔티티
- domain/MenuStatus : 메뉴의 상태값을 관리하기 위한 Enum class
- tables/MenuTable : 메뉴를 테이블 처럼 관리하기 위한 Map을 저장

JSeHoone committed:
feat : Loader와 Service 추가

- Loader/ : InMemoryData에 저장되어있는 Menu데이터를 Table에 저장하기 위함
- Service/ : 비즈니스 로직을 수행하는 패키지

JSeHoone committed:
refactor : service로 받아오는 데이터로 처리하도록 로직 수정

JSeHoone committed:
refactor : ProductMenu로 진입시 ProductLoader로 동기화 하기

JSeHoone committed:
refactor : 데이터 유지를 위한 싱글톤 패턴으로 수정

- 기존에 ProductMenu로 돌아왔을때, syncMenuTable이 초기화 되는 문제 발생
- 이를 해결하기 위해 initFlag를 이용 + 싱글톤 패턴으로 데이터가 유지되도록 수정

JSeHoone committed:
refactor : Enum타입의 MenuStatus를 한국어로 변경

- 기존에 Enum의 name이 영어로 설정 되어있던 부분을 한국어로 변경

JSeHoone committed:
refactor : 기존 menuBoard의 이름이 출력되는 부분에 상태값도 출력이 되도록 수정

JSeHoone committed:
refactor : MenuEntity에 Setter추가

- 기존에는 Setter가 필요 없을 줄 알았는데, 업데이트 기능을 구현하는 과정에서 필요하여 추가함

JSeHoone committed:
refactor : 싱글톤과 MenuStatus Enum 변경으로 인한 수정

- MenuManagerView는 싱글톤으로 변경하면서 코드 수정
- ProductMenuLoader는 Enum 이름 변경으로 인한 코드 수정

JSeHoone committed:
feat : MenuBoardSettingView에서 선택 페이지로 이동하는 로직 추가

- 1번 선택 시 가격설정 페이지로 이동
- 2번 선택 시 상태 변경 페이지로 이동

JSeHoone committed:
feat : 메뉴 가격 설정 페이지와 메뉴 상태값 변경 페이지 추가

- MenuPriceSettingView : 메뉴 가격 설정을 담당하는 페이지
- MenuStatusSettingView : 메뉴의 상태 값을 담당하는 페이지

JSeHoone committed:
feat : 메뉴 얻어오는 로직 수정 및 업데이트 로직 추가

- getMenuList() : 기존 전체를 가져오는 로직에서 삭제 상태는 안가져오도록 수정
- updateMenuPrice() : 메뉴의 가격을 변경하는 로직
- updateMenuStatus() : 메뉴의 상태값을 변경하는 로직

JSeHoone committed:
fix : 오타 수정

- '1'이라는 값이 들어가있었음.

JSeHoone committed:
feat : 신메뉴 등록, 메뉴 삭제 페이지로 이동 로직 추가

- MenuRegisterView는 신메뉴 등록 페이지로 이동
- MenuDeleteView는 메뉴 삭제 페이지로 이동

JSeHoone committed:
refactor : currentId 필드를 추가

- currentId 필드를 추가하여 새로 메뉴가 추가될 때, 해당 값으로 ID값을 넣어주도록 수정

JSeHoone committed:
feat : MenuDeleteView, MenuRegisterView 추가

- MenuRegisterView : 신메뉴를 등록하는 페이지
- MenuDeleteView : 메뉴판에 있는 메뉴를 삭제하는 페이지

JSeHoone committed:
refactor : 상태값 옵션에 삭제가 보이지 않도록 로직 변경

- 삭제는 삭제 페이지에서만 가능하도록 하기위해서 해당 로직을 추가

JSeHoone committed:
feat : 신메뉴 등록 및 메뉴 삭제 로직 추가

- registerNewMenu() : 신메뉴를 등록해주는 메서드
- removeMenu() : 기존 메뉴를 삭제해주는 메서드

JSeHoone committed:
fix : 없는 메뉴를 선택한 경우 오류 처리

- updateMenuPrice : 없는 메뉴의 가격을 설정하려는 경우
- updateMenuStatus : 없는 메뉴의 상태값을 설정하려는 경우
- removeMenu : 없는 메뉴를 삭제하려는 경우

JSeHoone committed:
feat : 이전 페이지로 돌아가는 기능 추가

- 이전 페이지는 ManagerDashboard로 돌아가는 것을 의미

JSeHoone committed:
feat : ProductMenu를 수행하기 위한 코드 추가

- getInstance()메서드를 호출하여 싱글톤으로 관리할 수 있도록 진행
- run()을 수행하여 ProductMenu 프로그램이 수행이 되도록 진행

dawwson commented on src/main/java/org/BackOffice/services/menu/domain/MenuStatus.java:
```diff
@@ -0,0 +1,13 @@
+    판매중(1),
+    품절(2),
+    신규(3),
+    삭제(4);
```
찾아보니 문법적으로 문제는 없으나 권장되지는 않는다고 하네요. 
`PENDING("결제대기")` -> 이런식으로 영문 enum + 한국어 매핑 필드를 같이 두는 절충안이 있다고 합니다.

dawwson commented on src/main/java/org/BackOffice/services/menu/view/MenuDeleteView.java:
```diff
@@ -0,0 +1,48 @@
+        }
+        System.out.println(deletedMenu.getMenuName() +"이(가) 삭제되었습니다.");
+        System.out.println("이전 페이지로 돌아갑니다.");
+        AddAndDeleteMenuView.view();
```
개인적으로는 한 개의 view가 다른 view로 네비게이션 해줄 책임이 있는가... 생각해봅니다.

dawwson reviewed (approved):
고생하셨습니다!

[좋은 점]
- domain, service, view 로 계층을 분리하신 점
- 메뉴 조회, 수정, 삭제 기능이 이해하기 쉽게 작성되었습니다.

[아쉬운 점]
- view 간의 결합도가 높아보입니다.

JSeHoone committed:
refactor : table 패키지 삭제

- repository 패키지 생성으로 인해 불필요한 table패키지 삭제

JSeHoone committed:
feat : repository 패키지 추가

- 기존 table 패키지의 역할을 대체
- 확장 가능하도록 interface구현
- Memory에서 관리하는 MemoryMenuRepository 구현

JSeHoone committed:
refactor : repository추가로 인한 로직 변경

- 기존 table패키지에서 menuTable에 직접 변경을 수행하는 로직에서
- MenuRepository를 이용하여 저장하는 방법으로 수정

JSeHoone committed:
refactor : 기존 enum 이름을 한국어에서 영어로 변경

- enum 타입의 이름을 영어로 변경
- 내부 값에 statusValue로 한국어 설정
- 해당 값을 관리하기 편하도록 statusCode값 추가

JSeHoone committed:
refactor : 기존 싱글톤으로 관리하는 로직 제거

- ProductMenu를 싱글톤 인스턴스로 관리하던 로직 제거
- 역할, 책임, 설계의도 작성

JSeHoone committed:
refactor : Repository 생성으로 인한 Service로직 수정

- 기존 table내 menuTable 의존성을 제거
- Repository를 통해 데이터 저장 변수에 대한 의존성 제거
- MenuRepository에서 제공하는 메서드로 데이터 핸들링 하도록 로직 수정

JSeHoone committed:
fix : update 로직 수정

- update시 save메서드 호출로 인해 동일 데이터가 축적됨
- findEntity를 그대로 반환해서 문제 해결

JSeHoone committed:
feat : View를 관리하기 위한 모듈 추가

- AbstractView : View의 추상 클래스
- View : View의 인터페이스
- ViewRouter : View의 의존성을 낮추기 위한 라우터

JSeHoone committed:
feat : MenuController 추가

- View와 Service의 의존성을 낮추기 위함.
- 업데이트 하는 화면에서는 Service의 getMenuList와 update메서드를 수행해야해서 이를 줄이기 위함

JSeHoone committed:
refactor : View 화면들 리팩토링

- AbstractView와 Controller를 적용한 코드로 수정
- AddAndDeleteMenuView 삭제
- MenuAddAndDeleteView 로 이름 변경

JSeHoone committed:
refactor : DashBoard내 ProductMenu 코드 수정

- 기존 instance 생성에서 생성자를 생성한 후 Show 메서드 수행하는 코드로 변경

JSeHoone committed:
feat : ProductMenuLoader로 메뉴 데이터 동기화

- InMemoryData에 있는 MENU데이터를 repository에 있는 Map데이터에 동기화하여 초기값 설정해주는 로직 추가

JSeHoone committed:
refactor : MenuEntity에 Id필드 추가

JSeHoone committed:
refactor : 불필요한 import제거

- 기존 main메서드로 테스틀르 수행할 때 사용한 import 제거

JSeHoone committed:
refactor : 기존 index로 id값을 출력하고 있던 오류 수정

- 기존 메뉴의 정보를 불러로기 위해서 index를 사용했는데,
- MenuEntity에 id를 추가하여 출력하도록 수정

JSeHoone committed:
fix : 라우터 누락 수정

- 2번 선택시, menuAddAndDelete 페이지로 넘어가지지 않는 문제 해결

JSeHoone committed:
refactor : find, delete메서드 Null허용 및 테스트 전용 메서드 추가

- findById시 Null이 가능하도록 수정
- delete시 Null이 가능하도록 수정
- 테스트시 menuStore를 초기화 하기 위한 메서드 추가

JSeHoone committed:
feat : MemoryMenuRepositoryTest 테스트 모듈 추가

- repository에 있는 각 메서드들의 테스트 케이스 추가
